doc/decisions/3_decided/internal_cache.md

```diff
@@ -172,12 +206,23 @@ Possible copy-on-write implementations are described in [another decision](../1_
 
 ## Decision
 
+Implement the copy-on-write approach.
+As we need COW for change tracking anyway, it makes sense to also use this approach for the internal cache.
+This also does not require any changes or restrictions to the current API.
+
+We keep a copy of all keys returned by the backends in memory.
+We use `ksBelow` to only return the keys the user requested on `kdbGet`.
+On `kdbSet` we use the cached in-memory keys to fill in the missing keys outside of `parentKey` that are needed for all the backends to save.
```

Comment:
I think much of this is actually "Rationale".

Also there should be more details here or above. It is not really clear which keys are COW, how they are kept, etc.

I assume, the only thing we change is that instead of returning keys from `backendData->keys` directly, we return a COW copy. So the only change would be in `backendsMerge` (possibly we need two versions or a flag parameter).

I would also say things differently for `kdbSet`. I'd say we take the internal cache as a base and add the user provided `KeySet` (below `parentKey`) on top thereby replacing any overlap.

Reply:
> I'd say we take the internal cache as a base and add the user provided KeySet (below parentKey) on top thereby replacing any overlap.

No, then the user could never delete keys. We can only use keys outside of `parentKey` from the cache.

Reply:
Ah that's true of course... I'd still not say "fill in the missing keys", IMO it sounds a bit weird and it's not really clear what is happening (e.g. which keys are "missing keys"). I made a concrete suggestion in another comment.